File: src/main/java/de/eugenbernwald/fretboardtrainer/model/GuitarString.java
package de.eugenbernwald.fretboardtrainer.model;

import java.util.*;

public class GuitarString {

    private Random random = new Random();
    private List<HashSet<Tone>> tones;

    GuitarString(Tone startTone) {

        this.tones = new ArrayList<>();

        Set<Note> allNotes = EnumSet.allOf(Note.class);
        Set<Shift> allShifts = EnumSet.allOf(Shift.class);
        int octave = startTone.getKey() / 12;

        for (int i = 0; i < 12; i++) {
            this.tones.add(new HashSet<>());
            for (Note note : allNotes) {
                for (Shift shift : allShifts) {
                    for (int j = -1; j <= 1; j++) {
                        Tone tone = new Tone(note, shift, octave+j);
                        if(Tone.key2Frequency(startTone.getKey() + i) == tone.getFrequency()){
                            this.tones.get(i).add(tone);
                        }
                    }



                }
            }
        }
    }

    public Tone getRandomToneOnFret(int num){
        Set<Tone> possibleTones = this.tones.get(num);
        return new ArrayList<>(possibleTones).get(this.random.nextInt(possibleTones.size()));
    }
}

File: src/main/java/de/eugenbernwald/fretboardtrainer/model/GuitarString.kt
package de.eugenbernwald.fretboardtrainer.model

import java.util.*

class GuitarString internal constructor(startTone: Tone) {

    private val random = Random()
    private val tones: MutableList<HashSet<Tone>>

    init {

        this.tones = ArrayList()

        val allNotes = EnumSet.allOf(Note::class.java)
        val allShifts = EnumSet.allOf(Shift::class.java)
        val octave = startTone.key / 12

        for (i in 0..11) {
            this.tones.add(HashSet())
            for (note in allNotes) {
                for (shift in allShifts) {
                    for (j in -1..1) {
                        val tone = Tone(note, shift, octave + j)
                        if (Tone.key2Frequency(startTone.key + i) == tone.frequency) {
                            this.tones[i].add(tone)
                        }
                    }


                }
            }
        }
    }

    fun getRandomToneOnFret(num: Int): Tone {
        val possibleTones = this.tones[num]
        return ArrayList(possibleTones)[this.random.nextInt(possibleTones.size)]
    }
}
